File: haskell/glue/src/Glue/Lib/Math/Logarithmic/Log.hs
module Glue.Lib.Math.Logarithmic.Log where

import Glue.Eval (Eval, throwError)
import Glue.Eval.Exception (wrongArgumentType)
import Glue.IR (IR (..))

-- Logarithm function (log base value)
-- Mirrors Haskell Glue.Lib.Math.Logarithmic.Log.log exactly
log :: IR Eval
log = NativeFunc logImpl

-- Logarithm function implementation
-- Mirrors Haskell Glue.Lib.Math.Logarithmic.Log.logImpl exactly
logImpl :: IR Eval -> Eval (IR Eval)
logImpl arg = pure $ NativeFunc (logWithBase arg)

logWithBase :: IR Eval -> IR Eval -> Eval (IR Eval)
logWithBase arg base = case (arg, base) of
    (Integer n, Integer b) -> pure $ Float (Prelude.logBase (fromIntegral b) (fromIntegral n))
    (Integer n, Float b) -> pure $ Float (Prelude.logBase b (fromIntegral n))
    (Float n, Integer b) -> pure $ Float (Prelude.logBase (fromIntegral b) n)
    (Float n, Float b) -> pure $ Float (Prelude.logBase b n)
    _ -> throwError $ wrongArgumentType ["number", "number"]

File: dart/glue/lib/src/lib/math/logarithmic/log.dart
import 'dart:math' as math;

import 'package:glue/src/eval.dart';
import 'package:glue/src/eval/exception.dart';
import 'package:glue/src/ir.dart';

/// Logarithm with arbitrary base function
/// Mirrors Haskell Glue.Lib.Math.Logarithmic.Log.log exactly
final Ir log = IrNativeFunc(logImpl);

/// Logarithm with arbitrary base function implementation
/// Mirrors Haskell Glue.Lib.Math.Logarithmic.Log.logImpl exactly
Eval<Ir> logImpl(Ir value) {
  return Eval.pure(IrNativeFunc(logWithBase(value)));
}

/// Helper function for base argument
/// Mirrors Haskell Glue.Lib.Math.Logarithmic.Log.logWithBase exactly
Eval<Ir> Function(Ir) logWithBase(Ir value) {
  return (Ir base) {
    return sequenceAll([eval(value), eval(base)]).flatMap((values) {
      return switch (values) {
        [IrInteger(value: final n), IrInteger(value: final b)] => Eval.pure(
          IrFloat(math.log(n.toDouble()) / math.log(b.toDouble())),
        ),
        [IrInteger(value: final n), IrFloat(value: final b)] => Eval.pure(
          IrFloat(math.log(n.toDouble()) / math.log(b)),
        ),
        [IrFloat(value: final n), IrInteger(value: final b)] => Eval.pure(
          IrFloat(math.log(n) / math.log(b.toDouble())),
        ),
        [IrFloat(value: final n), IrFloat(value: final b)] => Eval.pure(
          IrFloat(math.log(n) / math.log(b)),
        ),
        _ => throwError(wrongArgumentType(['number', 'number'])),
      };
    });
  };
}
